### ISSUE TITLE
BioCypher scalability

### ISSUE THREAD

**Issue by marco-brandizi** [biocypher/biocypher#295]
Hi all,

apologies if this isn't appropriate for an issue, couldn't find another place.

I've recently discovered BioCypher from your paper and it's very interesting as a potential build tool for the plant biology pipeline we manage in my team (KnetMiner), so I started looking into it better.

A first thing I'd like to understand is if BioCypher has scalability options. Namely, if it can exploit HPC cluster frameworks, such as LSF or SLURM, or if it can leverage cloud systems, such as Docker Swarm.

In particular, I'm thinking of building a knowledge graph from many data sources and running importers/mappers in parallel.

It seems BioCypher hasn't native support for this, but maybe there are ways to arrange scripts that could use these systems?

A similar question is if BioCypher allows for building a knowledge graph in a progressive way, in a way similar to workflow systems such as SnakeMake, eg, if a KG is built from a pipeline like: 

```
dataSrc1 -> dataSrc2 -> dataSrc3 -> KG
dataSrc4 ->-----------------------/
```

Is there a way to, say, start the whole pipeline, seeing it interrupted when dataSrc3 was imported and dataSrc2 was produced, and then, when I restart the workflow, it only produces dataSrc3 and imports it? Or similarly, could I run it partially when, for example, only dataSrc4 has changed? Is there a way to integrate BioCypher with workflow systems (like SnakeMake) that can do that (and support clusters)? 

Hope it makes sense!

Thanks in advance,
Marco.

**Comment by slobentanzer** [comment]
Hi @marco-brandizi,
Thanks for your request, this is absolutely the right place to ask. :)

You are right in assuming that there are no significant parallelisation functions in BioCypher. All your requests make perfect sense, and most are on our radar, some even on the roadmap. The main reasons they are not yet established is that it is very hard to automate prioritisation between resources (e.g., conflict handling), and that for most datasets it was not really necessary to run BioCypher highly parallel, as most builds are a one-time investment and biomedical datasets tend to be on the smaller side.

One of the next major steps of BioCypher development is the establishment of a higher-level pipeline system to automate the combination of resources. I'd be more than happy to include you and your use case in these developments, as I think it is always better to have a practical application for the development targets.

That said, I have already done some of the parallel scripts that you mentioned, but that only works if your resources are independent. E.g., if one type of node pulls information from multiple resources, this implementation is non-trivial. What may be of note is that the folks from the IGVF project (you can find the link on our component board) are successfully using pypy3 to speed up the pipeline runtime for their quite extensive dataset.

But I totally agree that, in general, cloud-enabled and snakemake-like orchestration would be great to have in the BioCypher core. Happy to talk more, and feel free to join our [Zulip channel](https://biocypher.zulipchat.com) for more informal chatting!
